libraries/state_history/log.cpp

```diff
@@ -288,21 +309,59 @@ void state_history_log::split_log() {
    index.open("w+b");
 }
 
+void state_history_log::store_entry(const chain::block_id_type& id, const chain::block_id_type& prev_id,
+                                    std::vector<char>&& data) {
+   block_num_type block_num = chain::block_header::num_from_id(id);
+   {
+      std::lock_guard<std::mutex> lk(mx);
+      entries.try_emplace(block_num, log_entry_type{id, prev_id, std::make_shared<std::vector<char>>(std::move(data))});
+   }
+   cv.notify_one();
+}
+
+void state_history_log::write_entry(const log_entry_type& entry) {
+   state_history_log_header header{.magic = ship_magic(ship_current_version), .block_id = entry.id};
+   auto [block_num, start_pos] = write_entry_header(header, entry.prev_id);
+   try {
+      mx.unlock();
+      this->write_payload(write_log, *entry.data);
+      mx.lock();
+      write_entry_position(header, start_pos, block_num);
+   } catch (...) {
+      write_log.close();
+      boost::filesystem::resize_file(write_log.get_file_path(), start_pos);
+      write_log.open("rb+");
+      throw;
+   }
+}
+
 state_history_traces_log::state_history_traces_log(const state_history_config& config)
-    : state_history_log("trace_history", config) {}
+    : state_history_log("trace_history", config) {
+}
 
-chain::bytes state_history_traces_log::get_log_entry(block_num_type block_num) {
+std::shared_ptr<std::vector<char>> state_history_traces_log::get_log_entry(block_num_type block_num) {
+   
+   std::lock_guard<std::mutex> lk(mx);
+   
+   auto itr = entries.find(block_num);
+   if (itr != entries.end()) {
+      std::vector<state_history::transaction_trace> traces;
+      auto ds = fc::datastream<const char*>(itr->second.data->data(), itr->second.data->size());
+      state_history::trace_converter::unpack(ds, traces);
+      return std::make_shared<std::vector<char>>(fc::raw::pack(traces));
```

Comment:
Do you think the pack can take long enough that it would be worth releasing the lock here?

Reply:
> Do you think the pack can take long enough that it would be worth releasing the lock here?

From the KV heavy contract, the traces have never been big enough to cause any significant impact, that's why I didn't try to make the lock more finer grained.